src/builder/QuarkBuilder.sol

```diff
@@ -377,19 +457,55 @@ contract QuarkBuilder {
         }
     }
 
+    function getWrapperCounterpartBalance(
+        string memory assetSymbol,
+        uint256 chainId,
+        Accounts.ChainAccounts[] memory chainAccountsList
+    ) internal pure returns (uint256) {
+        if (TokenWrapper.hasWrapperContract(chainId, assetSymbol)) {
+            // Add counterpart balance to balanceOnChain
+            return Accounts.getBalanceOnChain(
+                TokenWrapper.getWrapperCounterpartSymbol(chainId, assetSymbol), chainId, chainAccountsList
+            );
+        }
+
+        return 0;
+    }
+
     function needsBridgedFunds(
         string memory assetSymbol,
         uint256 amount,
         uint256 chainId,
         Accounts.ChainAccounts[] memory chainAccountsList,
         PaymentInfo.Payment memory payment
     ) internal pure returns (bool) {
+        uint256 balanceOnChain = Accounts.getBalanceOnChain(assetSymbol, chainId, chainAccountsList);
         // If action is paid for with tokens and the payment token is the transfer token, then add the payment max cost for the target chain to the amount needed
         uint256 amountNeededOnDstChain = amount;
         if (payment.isToken && Strings.stringEqIgnoreCase(payment.currency, assetSymbol)) {
             amountNeededOnDstChain += PaymentInfo.findMaxCost(payment, chainId);
         }
-        return Accounts.getBalanceOnChain(assetSymbol, chainId, chainAccountsList) < amountNeededOnDstChain;
+
+        // If has wrap token counterpart, try to wrap/unwrap first before attempt to bridge
+        uint256 wrapperCounterpartBalance = getWrapperCounterpartBalance(assetSymbol, chainId, chainAccountsList);
+        // Substract max cost if the wrapper counter part is the payment token for wrap/unwrap action
+        if (
+            payment.isToken
+                && Strings.stringEqIgnoreCase(
+                    payment.currency, TokenWrapper.getWrapperCounterpartSymbol(chainId, assetSymbol)
+                )
+        ) {
+            uint256 maxCost = PaymentInfo.findMaxCost(payment, chainId);
+            if (wrapperCounterpartBalance >= maxCost) {
+                wrapperCounterpartBalance -= maxCost;
+            } else {
+                // Can't afford to wrap/unwrap == can't use that balance
+                wrapperCounterpartBalance = 0;
+            }
+        }
+        balanceOnChain += wrapperCounterpartBalance;
```

Comment:
Ok. 👍 
I have created a `Math.sol` to handle such case, and realized not only this will have annoying `if` checker for substracting, so put it in generic math lib for other similar situations to use as well. 

For checking payment isToken,  part of me is leaning to as-is, as it's identical to the one level above `if` for the TransferIntent payment token check.  If we really wnat to modularize this into `available balance`  we can go in a bigger refactor route that also refactor the logics above (handling with intent token).  But unsure if we want to include those refactor changes in this PR as well haha.  (So reached to the only moving out the Math parts for this PR).